homeassistant/components/openplantbook/__init__.py

```diff
@@ -0,0 +1,97 @@
+"""The OpenPlantBook integration."""
+import logging
+
+import voluptuous as vol
+
+from homeassistant import exceptions
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import DOMAIN, PLANTBOOK_BASEURL
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the OpenPlantBook component."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up OpenPlantBook from a config entry."""
+
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {}
+
+    hass.data[DOMAIN]["PLANTBOOK_TOKEN"] = entry.data.get("token")
+    hass.data[DOMAIN]["API"] = OpenPlantBookApi(hass)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    hass.data.pop(DOMAIN)
+
+    return True
+
+
+class OpenPlantBookApi:
+    """Fetches data from the OpenPlantbook API."""
+
+    def __init__(self, hass):
+        """Initialize."""
+        self.hass = hass
+
+    async def get_plantbook_token(self, client_id, secret):
+        """Get the token from the openplantbook API."""
+        if not client_id or not secret:
+            return False
+        token = None
+        url = f"{PLANTBOOK_BASEURL}/token/"
+        data = {
+            "grant_type": "client_credentials",
+            "client_id": client_id,
+            "client_secret": secret,
+        }
+        try:
+            session = async_get_clientsession(self.hass)
+            async with session.post(url, data=data) as result:
+                token = await result.json()
+                if token.get("access_token"):
+                    _LOGGER.debug("Got token from %s", url)
+                    return token
+
+                raise InvalidAuth
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.error("Unable to connect to openplantbook")
+            raise
+
+    async def get_plantbook_data(self, species):
+        """Get information about the plant from the openplantbook API."""
+        if not self.hass.data[DOMAIN]["PLANTBOOK_TOKEN"]:
+            _LOGGER.debug("No plantbook token")
+            return
+        url = f"{PLANTBOOK_BASEURL}/plant/detail/{species}"
+        headers = {
+            "Authorization": f"Bearer {self.hass.data[DOMAIN]['PLANTBOOK_TOKEN'].get('access_token')}"
+        }
+        try:
+            session = async_get_clientsession(self.hass)
+            async with session.get(url, headers=headers) as result:
```

Comment:
Communication to external devices or services must be wrapped in an external Python library hosted on pypi.

See: https://developers.home-assistant.io/docs/development_checklist/
And: https://developers.home-assistant.io/docs/creating_component_code_review/#4-communication-with-devicesservices with link to tutorial how to get started with that.

Reply:
Thanks for the review.  First core component I am testing out, so this is a learning process...

Reply:
Great! I'm not sure what the value is of an integration without entities. What kind of entities should the Plant component expose and how would that work without OpenPlantbook as you describe should be possible?

Reply:
I agreee that it might look a bit strange, but I could not find a better way to handle this.
The Plant component accepts a lot of max/min values (max_moisture, min_moisture, max_temperature etc.)
https://www.home-assistant.io/integrations/plant/

Currently these have to be set manually for each plant in configuration.yaml.

The OpenPlantbook contains data for all of these settings for a lot of different plant species, so all you need to set for each plant is the name of the species, and the rest can be fetched automatically from the API.  But since the API requires authentication I need a way to configure and store the `client_id` and `secret` in HA to access the API.
I considered doing this directly in the Plant component, but that would require either changes to the way the Plant component is set up to allow setting the `client_id` and `secret` in `configuration.yaml`, or a big rewrite to set up each plant using config_flow, but with so many settings for each plant I think that would be pretty complicated.



So in a follow up PR, I am going to add some more config settings to the Plant component (including `species:`), and also a function that checks if this component (Openplantbook) is set up, and if so - fetch the max/min-values from the API.
The idea is discussed and explained in more detail in this post (and thread):
https://community.home-assistant.io/t/cloud-plant-db-with-api-for-plantcard/225207/2?u=olen

Without Openplantbook, the Plant component will continue to work exactly as today, but with a few more optional config settings.
With Openplantbook, you can skip a lot of the config settings of each plant, and they will be filled in automatically from the API.

Reply:
Hmm...I see your challenge. This requires some architectural discussion I think, let's wait and gather some opinions.

Reply:
I'm going to fix the tests etc. in the mean time.

Reply:
Btw. I am also slightly confused whether I am allowed to add more config options (e.g `species` and `image`) to the Plant component without converting the whole component to config flow.
As mentioned setting up a `plant` requires quite a lot of config lines with all the sensors etc. which might make a config flow a bit complicated.  And if a config flow is required, it should validate that the configured sensors exist and provides reasonable data etc., so that is pretty heavy.

Reply:
Does it even make sense to reuse Plant integration? Seems to me all that does is group entities and check some min/max values. If you expose entities in this integration with one device per plant it's automatically grouped. Also min/max check isn't that complex to add. You can then taylor the entities here to fit API of OpenPlantbook.

Reply:
I'm not sure I understand what you mean.  How would I go ahead and group a bunch of (possibly unrelated) sensors under a new "plantbook-device"?
All the sensors measuring temperature, conductivity etc. are today individual entities without any internal relationship.

Also, "plants" are used in a lot of lovelace cards, automations etc. so I don't want to lose any of that functionality.

Reply:
I may not fully understand the problem. When you use OpenPlantbook, isn't all the data coming from OpenPlantbook? I you create them as entities within this integration you have control over to which device they belong. Or is there no logic when reading from OpenPlantbook what the relation is between measurements?

Reply:
Maybe question would be should Plant use OpenPlantbook or should OpenPlantbook use Plant? I think I lack knowledge of both to effectively discuss this.

Reply:
This integration is only to support a simpler configuration of a Plant.
Today, you "must" find the appropriate values and configure this for each plant manually:
```
my_plant:
  max_moisture: 100
  min_moisture: 1
  max_conductivity: 100
  min_conductivity: 1
  max_brightness: 100
  min_brightness: 1
  max_temperature: 100
  min_temperature: 1
  sensors:
    moisture: sensor.xxx
    conductivity: sensor.yyy
    temperature: sensor.zzz
    brightness: sensor.aaa

```
The only thing openplantbook provides are the values for max_* and min_* based on the species of the plant pluss a link to a image of the plant species.

So the next step if/when this integration is accepted is to patch the plant integration to allow for a simplified config for each plant:
```
my_plant:
  species: capsicum annuum
  sensors:
    moisture: sensor.xxx
    conductivity: sensor.yyy
    temperature: sensor.zzz
    brightness: sensor.aaa

```
And when `my_plant` is set up (on startup of HA) all the values for max_* and min_* is fetched from the Openplantbook API and added to the plants config dynamically. The data is only pulled once - on setup of a new plant entity.

So it has nothing to do with the sensors configured for the plant.

Reply:
I start to understand what you are trying to achieve :)

Reply:
> I start to understand what you are trying to achieve :)

That is a good first step!  ;)
I added a couple of services to the integration, so it can be used by other integrations to search and get data from the API.
Hopefully that makes it more likely to get added to HA.

Reply:
And. btw.  I don't understand how to fix the remaining check that fails:
> codecov/patch — 90.52% of diff hit (target 96.18%)

I am not sure how many other tests I can add that will make any sense.

I am also currently working on documentation of the integration.